src/background/http-instrument.ts

```diff
@@ -1,687 +1,693 @@
 
-function logWithResponseBody(respEvent, update) {
-  // log with response body from an 'http-on-examine(-cached)?-response' event
-  const newListener = new TracingListener();
-  respEvent.subject.QueryInterface(Ci.nsITraceableChannel);
-  newListener.originalListener = respEvent.subject.setNewListener(newListener);
-  newListener.promiseDone
-    .then(
-      function() {
-        const respBody = newListener.responseBody; // get response body as a string
-        const bodyBytes = converter.convertToByteArray(respBody); // convert to bytes
-        cryptoHash.init(cryptoHash.MD5);
-        cryptoHash.update(bodyBytes, bodyBytes.length);
-        const contentHash = binaryHashtoHex(cryptoHash.finish(false));
-        update.content_hash = contentHash;
-        loggingDB.saveContent(
-          loggingDB.escapeString(respBody),
-          loggingDB.escapeString(contentHash),
-        );
-        loggingDB.saveRecord("http_responses", update);
+        if (oldNotifications) {
+          return oldNotifications.getInterface(iid);
+        } else {
+          throw Cr.NS_ERROR_NO_INTERFACE;
+        }
       },
-      function(aReason) {
-        loggingDB.logError(
-          "Unable to retrieve response body." + JSON.stringify(aReason),
+
+      asyncOnChannelRedirect(oldChannel, newChannel, flags, callback) {
+        const isTemporary = !!(flags & Ci.nsIChannelEventSink.REDIRECT_TEMPORARY);
+        const isPermanent = !!(flags & Ci.nsIChannelEventSink.REDIRECT_PERMANENT);
+        const isInternal = !!(flags & Ci.nsIChannelEventSink.REDIRECT_INTERNAL);
+        const isSTSUpgrade = !!(
+          flags & Ci.nsIChannelEventSink.REDIRECT_STS_UPGRADE
         );
-        update.content_hash = "<error>";
-        loggingDB.saveRecord("http_responses", update);
+
+        newChannel.QueryInterface(Ci.nsIHttpChannel);
+
+        const httpRedirect: HttpRedirect = {
+          crawl_id: crawlID,
+          old_channel_id: oldChannel.channelId,
+          new_channel_id: newChannel.channelId,
+          is_temporary: isTemporary,
+          is_permanent: isPermanent,
+          is_internal: isInternal,
+          is_sts_upgrade: isSTSUpgrade,
+          time_stamp: new Date().toISOString(),
+        };
+        this.dataReceiver.saveRecord("http_redirects", httpRedirect);
+
+        if (oldEventSink) {
+          oldEventSink.asyncOnChannelRedirect(
+            oldChannel,
+            newChannel,
+            flags,
+            callback,
+          );
+        } else {
+          callback.onRedirectVerifyCallback(Cr.NS_OK);
+        }
       },
-    )
-    .catch(function(aCatch) {
-      loggingDB.logError(
+    };
+    */
+
+    const httpRedirect: HttpRedirect = {
+      crawl_id: crawlID,
+      old_channel_id: null, // previously: oldChannel.channelId, TODO: Refactor to corresponding webext logic or discard
+      new_channel_id: details.requestId, // previously: newChannel.channelId, TODO: Refactor to corresponding webext logic or discard
```

Comment:
From reading the docs, I think this should be assigned to `old_channel_id`. The `details` object has a `url` and a `redirectUrl` property, which makes me think the listener, and thus (`requestId`) applies to the original channel prior to a redirection.

My guess is the best you'll be able to do is pull the URLs. This is unfortunate because it means we'll have to build redirect chains using non-unique identifiers. URLs are non-unique because multiple resources can redirect to the same `redirectUrl`. If this is the case you'll want to save as much context as possible to disambiguate (e.g., parentFrameId, tabId, etc).

The `redirectUrl` is [pulled from `newChannel.originalURI.spec`](https://searchfox.org/mozilla-central/source/toolkit/modules/addons/WebRequest.jsm#919). Note that the [`originalURI` channel property](https://searchfox.org/mozilla-central/source/netwerk/base/nsIChannel.idl#46) may not be the same as the URI that is passed to the web extension in the resulting request, so linking channels during a redirect may require additional heuristics such as resolving relative locations and so on (depending on what's stored in `originalURI`.

Ideally WebExtensions would expose `redirectId` in addition to `redirectUrl`.